### PR TITLE
[Partial] Beanstalk MQ support for magneticod

### PR DESCRIPTION
**Pull request by skobkin**
In this PR I implemented simple database engine for PostgreSQL.
It implements `pkg/persistence/interface.go` partially only for `magneticod` part.

It uses (and slightly refactors) the same format as `stdout` engine.

### CODE REVIEW

File: pkg/persistence/interface.go [hunk 1]
```diff
@@ -88,6 +89,12 @@ type TorrentMetadata struct {
 	Relevance    float64 `json:"relevance"`
 }
 
+type SimpleTorrentSummary struct {
```
[the remark this answers is not in the document]
Comment (skobkin) [reply pkg/persistence/interface.go]: Other exported structures does not have comments. Do I have to add this comment?

File: pkg/persistence/beanstalkd.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	"go.uber.org/zap"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/iwanbk/gobeanstalk"
+	"github.com/pkg/errors"
+)
+
+var beanstalkdWriteOnlyError = errors.New("This dummy storage engine (\"beanstalkd\") is write-only")
```
Comment (skobkin) [inline pkg/persistence/beanstalkd.go]: Probably we need to make common error for this type of "not implemented" or "magneticod only" errors like I did with `SimpleTorrentSummary` in `interface.go`.

**Comment by lruggieri** [comment]
I guess every new DB should also have a Docker image and the main magneticod image should depend on it

**Comment by skobkin** [comment]
@lruggieri I don't think so. At least with current magnetico architecture it would be rather complicated to separate all engines from the *core* tool.

Do you have understanding how to implement your scheme?

**Comment by lruggieri** [comment]
> @lruggieri I don't think so. At least with current magnetico architecture it would be rather complicated to separate all engines from the _core_ tool.
> 
> Do you have understanding how to implement your scheme?

If you want to make this the main DB for the project I'd say to dockerize the DB itself, sharing the data folder and then put a depends_on POSTGRES_SERVICE in magneticod service.

If, instead, the purpose is letting the user choose between several DB systems...well, in that case, I would still dockerize every DB and maybe not putting the dependency in the compose file. In that case there should however be a nice explanation in the README about how the container of the DB of choice should be run before starting the main program.

**Comment by skobkin** [comment]
@lruggieri Huh. I thought that you're talking about magneticod code itself.

If you talking about **how to run it in the Docker** then this PR is **not about it** at all.

BTW I still don't fully understand what's you're trying to say.

> If you want to make *this* the main DB for the project I'd say to dockerize the DB itself

What's "this"? If "this" is the magneticod then you already have two `*.Dockerfile` in the root of this repository to build `magneticod` and `magneticow`.

If "this" is Beanstalkd then this PR is not about how to run Beanstalkd. It's just implementing *persistence*/storage engine for `magneticod` to be able to put every discovered torrent as new job in Beanstalkd.

How to run any storage back-end is **your** choice. This PR is only about supporting one of the storages.

**Comment by lruggieri** [comment]
@skobkin you are right, reading "In this PR I implemented simple database engine for PostgreSQL." in the first message I actually thought this PR was actually implementing the Postgres DB interface.
My bad, I should gone through the commit changes first. 

This is actually a queue to put SimpleTorrentSummary objects there to be used, right? I still don't fully see the purpose though. Who should use those queued jobs?

**Comment by skobkin** [comment]
> This is actually a queue to put SimpleTorrentSummary objects there to be used, right?

Yes. I've just used already existing `out` structure from `stdout` engine but renamed it for common use.

> Who should use those queued jobs?

Anyone who want to process indexed torrents or save them in their own database schema.

For example using `beanstalk` engine you can implement #223 in your application logic not relying on magneticod.

Also you can implement almost instant torrent availability notification system for your users. To implement something like these you'll need to use triggers or perpetual polling from the database.

So basically it's up to your imagination.

File: pkg/persistence/beanstalkd.go (hunk 2 above)
Comment (lruggieri) [reply pkg/persistence/beanstalkd.go]: we could create a WriteOnly interface that implements all the functions which read on data (GetNumberOfTorrents, get Torrent, Getdata etc.). Then Beanstalk (or any other kind of write-only db like StdOut) could build on it implementing the remaining "write" function necessary to get to a full Database implementation

Comment (skobkin) [reply pkg/persistence/beanstalkd.go]: @lruggieri Totally agree with this idea. But I'm not ready to do this refactoring myself.

Comment (lruggieri) [reply pkg/persistence/beanstalkd.go]: @skobkin I did a PR on your own branch, implementing this kind of interface. See if you like it.

**Comment by lruggieri** [comment]
@skobkin 
I was thinking that magnetico should probably not implement internally any other DB interface other than SQLite, and just adding a boolean option for enabling the queue or not.
Then the whole queue system should be properly dockerized (queue included: installation, ports exporting, settings and stuff) and any user who wants to implement an external DB (like me with Elasticsearch + Kibana) would just have to connect to the queue on the container.

In this instance the queue should not even be considered a Database anymore.

File: pkg/persistence/beanstalkd.go (hunk 2 above)
Comment (skobkin) [reply pkg/persistence/beanstalkd.go]: @lruggieri I saw it. Looks good *for me*. But it's would be better to propose such changes to the upstream and then if it's approved I can rebase my PR on them.

Proposing both in my PR would not be good for segregation of proposed changes.
